src/index/VocabularyGenerator.cpp

```diff
@@ -75,86 +70,166 @@ size_t mergeVocabulary(const std::string& basename, size_t numFiles,
     }
   }
 
-  // keep track of the last seen word to correctly handle duplicates
-  std::string lastWritten = "";
-  // the number of words we have written. This also is the global Id of the next
-  // word we see, unless it is is equal to the previous word
-  size_t totalWritten = 0;
-  bool firstLangPredSeen = false;
-  *langPredLowerBound = 0;
-  *langPredUpperBound = 0;
+  std::vector<QueueWord> sortedBuffer;
+  sortedBuffer.reserve(_bufferSize + 2);
+
+  std::future<void> writeFuture;
 
   // start k-way merge
   while (!queue.empty()) {
-    auto top = queue.top();
+    sortedBuffer.push_back(std::move(queue.top()));
     queue.pop();
+    auto i = sortedBuffer.back()._partialFileId;
+
+    if (sortedBuffer.size() >= _bufferSize) {
+      auto task = [this, buf = std::move(sortedBuffer)]() {
+        this->writeQueueWordsToIdVec(buf);
+      };
+      sortedBuffer.clear();
+      sortedBuffer.reserve(_bufferSize + 2);
```

Comment:
Can you add a comment about the `+2`

Reply:
There is no reason, I was ruling out possible performance pitfalls at one point, I am going to remove it.